accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+    *A null-check pattern also modifies the matched value type of the subpattern
+    during type inference. This means that the subpattern usually has a
+    non-nullable type already, so step 2 above rarely comes into play. For
+    example:*
 
-*   **Variable pattern:** An object space whose type is the variable's type
-    (which might be inferred).
+    ```dart
+    typedef NullableObject = Object?;
 
-*   **Literal or constant pattern:** These are handled specially depending on
-    the constant's type:
+    test(Object? obj) => switch (obj) {
+      case _?:
+      case null:
+    };
+    ```
 
-    *   We treat `bool` like a sealed supertype with subtypes `true` and
-        `false`. The Boolean constants `true` and `false` are lifted to object
-        patterns of those subtypes.
+    *Here, the inferred type of the inner `_` pattern is `Object` and thus its
+    lifted space is also `Object`. But if the inner pattern happens to be
+    nullable, then step 2 can be involved:*
 
-    *   Likewise, we treat enum types as if the enum type was a sealed supertype
-        and each value was a singleton instance of a unique subtype for that
-        value. Then an enum constant is lifted to an object pattern of the
-        appropriate subtype.
+    ```dart
+    typedef NullableObject = Object?;
 
-        In the previous card example, we treat `Suit` as a sealed supertype with
-        subtypes `club`, `diamond`, `heart` and `spade`.
+    test(Object? obj) => switch (obj) {
+      case Object? _?:
+    };
+    ```
 
-        (We are *not* proposing adding Java-style enum value subtypes to the
-        Dart language. This is just a way to model enums for exhaustiveness
-        checking.)
+    *Here, the subspace expands to `Object|Null` and the space for the
+    surrounding null-check pattern yields just `Object`.*
 
-    *   We lift other constants to an object pattern whose type is a synthesized
-        subtype of the constant's type based on the constant's identity. Each
-        unique value of the constant's type is a singleton instance of its own
-        type, and the constant's type behaves like an *unsealed* supertype. Two
-        constants have the same synthesized subtype if they are identical
-        values.
+*   **Null-assert pattern:** A union of the lifted space union of the subpattern
+    and a space with type `Null`.
 
-        This means you don't get exhaustiveness checking for constants, but do
-        get reachability checking:
+    *As with cast patterns, a null-assert pattern "matches" `null` by throwing
+    an exception, which is sufficient for exhaustiveness.*
 
-        ```dart
-        String s = ...
-        switch (s) {
-          case 'a string': ...
-          case 'a string': ...
-        }
-        ```
+*   **Constant pattern:** A space whose type is the type of the constant and
+    with a constant restriction for the given constant value.
```

Comment:
(Do we require the type of a constant pattern to have primitive equality? If not, should this only apply if it does?)

Reply:
Good question. No, we don't require constants that appear in constant patterns to have primitive equality. In practice, the only constants that actually usefully participate in exhaustiveness checking are bools and enum values, which do have primitive equality. So it mostly doesn't matter.

But it does potentially matter for reachability. Filed: https://github.com/dart-lang/sdk/issues/51873

Updated the spec here to match what that issue suggests.